recipes/FileUploadReactHooks/src/Main.purs

```diff
@@ -0,0 +1,50 @@
+module FileUploadReactHooks.Main where
+
+import Prelude
+import Data.Foldable (traverse_)
+import Data.Maybe (Maybe(..))
+import Data.Traversable (traverse)
+import Effect (Effect)
+import Effect.Exception (throw)
+import React.Basic.DOM (render)
+import React.Basic.DOM as R
+import React.Basic.DOM.Events (currentTarget)
+import React.Basic.Events (handler)
+import React.Basic.Hooks (Component, component, fragment, useState', (/\))
+import React.Basic.Hooks as React
+import Web.File.File as File
+import Web.File.FileList as FileList
+import Web.HTML (window)
+import Web.HTML.HTMLDocument (body)
+import Web.HTML.HTMLElement (toElement)
+import Web.HTML.HTMLInputElement as HTMLInputElement
+import Web.HTML.Window (document)
+
+main :: Effect Unit
+main = do
+  body <- body =<< document =<< window
+  case body of
+    Nothing -> throw "Could not find body."
+    Just b -> do
+      fileUploadComponent <- mkFileUploadComponent
+      render (fileUploadComponent {}) (toElement b)
+
+mkFileUploadComponent :: Component {}
+mkFileUploadComponent = do
+  component "FileUploadComponent" \_ -> React.do
+    fileList /\ setFileList <- useState' []
+    let
+      handleChange t =
+        join <$> traverse HTMLInputElement.files (HTMLInputElement.fromEventTarget t)
+          >>= traverse_ (FileList.items >>> map File.name >>> setFileList)
```

Comment:
I think this could be improved, at least readability-wise 🤔

Reply:
Yeah...

Why not move the `join` into its own block followed by a `do`?
```
handleChange t = join do
  -- code
```
If `traverse` was replaced with `for`, would that make it easier to read?

Reply:
> If `traverse` was replaced with `for`, would that make it easier to read?

Possibly? But I don't know if there's any way of removing the `join` (aside from explicit pattern matching), since
```haskell
traverse files (fromEventTarget t)
```
gives me an `Effect (Maybe (Maybe FileList))`.

I could do this:
```haskell
handleChange t =
  traverse HTMLInputElement.files (HTMLInputElement.fromEventTarget t)
    >>= join
    >>> traverse_ (FileList.items >>> map File.name >>> setFileList)
```
but maybe better to be more explicit
```haskell
      handleChange t = case HTMLInputElement.fromEventTarget t of
        Nothing -> pure unit
        Just fileInput -> do
          maybeFiles <- (HTMLInputElement.files fileInput)
          for_ maybeFiles
            (FileList.items >>> map File.name >>> setFileList)
```
?

I think I like the latter, I feel like if I wrote it like I had it originally I would have trouble understanding it if I came back to it like a week from now 😅.